python/ray/_private/async_utils.py

```diff
@@ -1,27 +1,86 @@
 #    distributed under the License is distributed on an "AS IS" BASIS,
 #    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
-
 # Modifications:
 # - Removed the dependency to `logwood`.
 # - Renamed `monitor_loop_lag.enable()` to just `enable_monitor_loop_lag()`.
 # - Miscellaneous changes to make it work with Ray.
 
-from typing import Callable, Optional
-import asyncio
-import asyncio.events
+
+async def async_call_with_retry(
+    f: Callable[[], Coroutine[Any, Any, Any]],
+    *,
+    exception_classes: Iterable[Type] = (),
+    max_attempts: int = 3,
+    initial_delay_s: float = 0.5,
+    max_backoff_s: float = 5,
+    on_exception: Optional[Callable[[Exception], None]] = None,
+) -> Any:
+    """
+    Helper utility retrying a provided async function with exponential backoff
+    policy.
+
+    NOTE: We're currently hard-coding 2 as an exponent base, meaning that
+          `initial_delay_s` will be doubled with every retry attempt up to
+          `max_backoff_s`.
+
+    Args:
+        c: Coroutine to be awaited on.
+        exception_classes: list of exception classes that should be retried (note that,
+                            all of the exceptions have to inherit from `BaseException`
+                            and could NOT extend `CancelledError`).
+        max_attempts: The maximum number of attempts to invoke.
+        initial_delay_s: Initial delay before the first retry
+        max_backoff_s: The maximum number of seconds to backoff.
+        on_exception: Callback to be invoked with an exception being retried.
+    """
+    assert all(
+        [issubclass(c, BaseException) for c in exception_classes]
+    ), "Provided exception classes have to inherit from BaseException"
+
+    assert not any(
+        [issubclass(c, CancelledError) for c in exception_classes]
+    ), "`CancelledError` should not be retried"
+
+    assert (
+        initial_delay_s >= 0
+    ), f"`initial_delay_s`, must be non-negative value (got {initial_delay_s}"
+    assert (
+        max_backoff_s >= 0
+    ), f"`max_backoff_s`, must be non-negative value (got {max_backoff_s}"
+    assert max_attempts >= 1, f"`max_attempts` must be >= 1 (got {max_attempts})"
+
+    for i in range(max_attempts):
+        try:
+            return await f()
+        except Exception as e:
+            is_retryable = any(
+                [isinstance(e, exc_class) for exc_class in exception_classes]
+            )
+            if is_retryable and i + 1 < max_attempts:
+                # Invoke callback if provided
+                if on_exception is not None:
+                    on_exception(e)
+
+                # Retry with (capped) exponential backoff
+                backoff = min((initial_delay_s * 2 ** (i + 1)), max_backoff_s)
+                await asyncio.sleep(backoff)
+            else:
+                raise e
```

Comment:
nit: wrap the exception with something saying it was retried X times or total duration